### PR TITLE
Change reductions inheritance to use sklearn BaseEstimator and mixins

### PR DESCRIPTION
**Pull request by romanlutz**
This is a first step in addressing #255 .

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -10,6 +10,11 @@
     an estimator through the `estimator` parameter. To use a pre-fitted
     estimator, pass `prefit=True`.
 * Rename arguments of `create_group_metric_set()` to match the dashboard
+* Remove `Reduction` base class for reductions methods and replace it with
+  `sklearn.base.BaseEstimator`, `sklearn.base.ClassifierMixin`, and
```
Comment (riedgar-ms) [inline CHANGES.md]: What functionality do these add, and should there be a smoke test to verify that such functionality really is present?

File: fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py [hunk 2]
```diff
@@ -22,7 +22,7 @@ def _mean_pred(X, hs, weights):
     return pred[weights.index].dot(weights)
 
 
-class ExponentiatedGradient(Reduction):
+class ExponentiatedGradient(BaseEstimator, ClassifierMixin):
     """An Estimator which implements the exponentiated gradient approach to reductions.
```
Comment (MiroDudik) [inline fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py]: This should be `BaseEstimator` and `MetaEstimatorMixin`. I don't mind having `ClassifierMixin` for now, but it will go away once we support `ExponentiatedGradient` with regression.

File: CHANGES.md (hunk 1 above)
Comment (adrinjalali) [reply CHANGES.md]: You can eventually pass the sklearn tests as whone [here](https://scikit-learn.org/stable/auto_examples/release_highlights/plot_release_highlights_0_22_0.html#checking-scikit-learn-compatibility-of-an-estimator).

``` python
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.utils.estimator_checks import parametrize_with_checks


@parametrize_with_checks([LogisticRegression, DecisionTreeRegressor])
def test_sklearn_compatible_estimator(estimator, check):
    check(estimator)
```

File: fairlearn/reductions/_grid_search/grid_search.py [hunk 3]
```diff
@@ -79,7 +79,7 @@ def accumulate_integer_grid(self, index, max_val):
                 self.accumulate_integer_grid(index + 1, max_val - abs(current_value))   # noqa: E501
 
 
-class GridSearch(Reduction):
+class GridSearch(BaseEstimator, ClassifierMixin, RegressorMixin):
     """Estimator to perform a grid search given a blackbox estimator algorithm.
```
Comment (MiroDudik) [inline fairlearn/reductions/_grid_search/grid_search.py]: This should be `BaseEstimator` and `MetaEstimatorMixin`.

As with `ExponentiatedGradient`, it would be to set `_estimator_type` and `score` functionality according to our base estimator... but happy to leave that for a separate PR.

File: CHANGES.md (hunk 1 above)
Comment (romanlutz) [reply CHANGES.md]: That would require a few more changes to work, right?

File: fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py (hunk 2 above)
Comment (romanlutz) [reply fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py]: I don't currently see `_estimator_type` and `score`, so I will leave them for another PR. Addressed your other comments, though!

File: CHANGES.md (hunk 1 above)
Comment (adrinjalali) [reply CHANGES.md]: Yes, I'm working on adding the test for ThresholdOptimizer, and once that's done, I guess we'd all have a better idea of what needs to be changed for it to work. In the mean time, this may help understanding what needs to be done:

https://fosdem.org/2020/schedule/event/python2020_scikit_learn_estimator/
